src/StaticLS/IDE/CodeActions/Parse.hs

```diff
@@ -0,0 +1,136 @@
+module StaticLS.IDE.CodeActions.Parse (
+  ActionableIssue (..),
+  Ignored (..),
+  KnownExtension (..),
+  actionableIssue,
+) where
+
+import Control.Applicative ((<|>))
+import Control.Monad (guard, (<=<))
+import Data.Foldable (asum, toList)
+import Data.Functor (($>))
+import Data.Text qualified as T
+import Language.Haskell.TH.LanguageExtensions (Extension)
+import Language.LSP.Protocol.Types qualified as LSP
+import Text.Regex.TDFA (getAllTextMatches, (=~), (=~~))
+
+-- This allows us to ignore diagnostics when comparing `ActionableIssue`s.
+-- Otherwise, we end up with duplicates with slightly different wording.
+newtype Ignored a = Ignored a deriving (Show)
+instance Eq (Ignored a) where (==) _ _ = True
+instance Ord (Ignored a) where compare _ _ = EQ
+
+data KnownExtension = KnownExtension {-# UNPACK #-} !Extension {-# UNPACK #-} !T.Text
+  deriving (Eq, Ord, Show)
+
+{-# NOINLINE knownExtensions #-}
+knownExtensions :: [KnownExtension]
+knownExtensions = (\ext -> KnownExtension ext $ T.pack $ show ext) <$> [minBound .. maxBound]
+
+data ActionableIssue
+  = RequiredExtension (Ignored LSP.Diagnostic) KnownExtension
+  | TypedHoleFits (Ignored LSP.Diagnostic) [T.Text]
+  | MissingFields (Ignored LSP.Diagnostic) T.Text (Maybe T.Text) [T.Text]
+  | MissingMethods (Ignored LSP.Diagnostic) [T.Text]
+  | MissingAssociatedType (Ignored LSP.Diagnostic) T.Text
+  | MissingCasses (Ignored LSP.Diagnostic) [T.Text]
+  deriving (Eq, Ord, Show)
+
+actionableIssue :: LSP.Diagnostic -> Maybe ActionableIssue
+actionableIssue diag =
```

Comment:
The parsing code in this module follows similar code in HLS. They literally do the same asinine thing for their quick fixes. That is, instead of using structured data that GHC would hypothetically provided, they parse the human-readable error messages. Go Haskell!